### DISCUSSION TITLE
Bots - APITriggers

### DISCUSSION THREAD

**Discussion by Yiakman** [Cotalker/documentation#170] [3. Ideas & Feature Requests] [answered]
Muchas de las implementaciones que se realizan por faas consisten en apis que procesan un request y con eso llaman una rutina en bots parametrizados. Podríamos abarcar varios de estos casos si es que existiese una forma de gatillar PB's mediante un post a una API. Cómo me lo imagino:

Los campos del body en JSON se traducen a la data que se le pasaría al bot, es decir, "$VALUE#data|name" sería equivalente a leer "req.body.data.name". El bot se ejecutaría cada vez que se realiza un post a un uri generado con el id o code de la rutina. Este caso tiene una particular sinergia con los embedded groups cuyas vistas ya tienen acceso a un jwt temporal de cotalker.

Existen dos modalidades para esta API, interna y externa. Interna requiere autentificación con un jwt de cotalker (o api key) y podría revisar los accessroles de este antes de ejecutar. Externa puede funcionar de forma abierta o bien se podrían generar y gestionar API keys desde el admin para estos bots.

**Comment by acedward** [comment, marked as answer]
No se si mal entendí - pero es posible configurar BOTS que no se gatillan por ningun trigger interno.
Y pueden ser invocados por su _id en una llamada de red (como un webhook)

Aquí desde Zapier capturamos un email desde un proveedor y gatillamos un BOT

![image](https://user-images.githubusercontent.com/3951895/117829612-f1d74480-b240-11eb-92b9-6047038bcdb0.png)

Con CustomCodeJS procesamos el email y convertimos su body en data estructurada

![image](https://user-images.githubusercontent.com/3951895/117830038-467abf80-b241-11eb-9846-e68c7931425b.png)

**Reply by Yiakman** [reply]
Claro, lo que digo es que haya un estándar de funcionamiento para ese webhook. Para el caso de que venga con data JSON en el body, pasar ese body como el campo data en el body del scheduler a ejecutar y se pueda leer directamente en COTlang

**Reply by acedward** [reply]
Así funciona exactamente!
En este ejemplo se puede leer $VALUE#email para leer el body del email

**Reply by Yiakman** [reply]
También una vez ejecutado el bot podría por defecto devolverse como respuesta al request el/los output de la rutina

**Reply by acedward** [reply]
El CCJS es solo un paso para poder convertir esa data no estructurada en estructurada.
CCJS lee los $VALUE#batata, como  value.batata en javascript (aun no liberamos la documentación - estamos aun probándolo)

**Reply by acedward** [reply]
@Yiakman también puedes leer la respuesta del bot! Pero tienes que hacer una segunda llamada - porque son async e..g, podrias programar una hora de ejecucción - pero te retorna el código para ser leeido

**Reply by acedward** [reply]
![image](https://user-images.githubusercontent.com/3951895/117831533-a6259a80-b242-11eb-98f0-3cae9b64c440.png)
 Estos son los campos retornados, hay que leerlos hasta que el status code diga ejecutado (el scheduler corre las cosas tan rápido como puede, pero nos ha pasado que hay una cola gigante de espera)

**Reply by Yiakman** [reply]
Cuáles son los campos para esa configuración?
Por defecto queda público ese endpoint?

**Reply by acedward** [reply]
```
POST /api/v1/bots/run/:id
Permissions admin-*-write 
body: any // es pasado directamente a tu bot para ser consumido con COTLang
```
